### DISCUSSION TITLE
SNES VI and mixed systems

### DISCUSSION THREAD

**Discussion by colinjcotter** [firedrakeproject/firedrake#3403] [Firedrake support] [answered]
I would like to set up a mixed system with VI constraints for one variable but not the other. 
Is it possible to specify that using Firedrake PETSc options?

**Comment by wence-** [comment, marked as answer]
You always have to pass in functions that represent the bounds in the full space, but you can put "unobtainable" bounds in the slots you don't want to constrain. See https://petsc.org/main/manualpages/SNES/SNESVISetVariableBounds/

So something like:

```python
import numpy as np
from firedrake.datatypes import ScalarType # can't remember where this one lives
MAX_REAL = np.finfo(ScalarType).max
# recreating code in petscmath.h
BOUND_POS_INF = MAX_REAL / 4
BOUND_NEG_INF = -BOUND_POS_INF
lbound = Function(W)
lbound.sub(0).interpolate(some_known_bound)
lbound.sub(1).interpolate(BOUND_NEG_INF)
... # same for upper bound

solver = ...
solver.solve(bounds=(lbound, ubound))
```

**Reply by danshapero** [reply]
This is my understanding of it too. I used `PETSc.INFINITY` and `NINFINITY` instead of `MAX_REAL` but same idea. I have some example code [here](https://shapero.xyz/posts/overland-flow/) if you want to see. I'd love to know if there's a way to use a VI solver for just the part of the system that needs it and other methods for the remainder.

**Reply by wence-** [reply]
Ah, I hadn't spotted that `PETSc.INFINITY` and `PETSc.NINFINITY` were exposed

**Reply by colinjcotter** [reply]
thanks!
